lightning-block-sync/src/convert.rs

```diff
@@ -541,4 +608,88 @@ pub(crate) mod tests {
 			Ok(txid) => assert_eq!(txid, target_txid),
 		}
 	}
+
+	/// TryInto<Transaction> can be used in two ways, first with plain hex response where data is the hex encoded transaction (e.g. as a result of getrawtransaction)
+	/// or as a JSON object where the hex encoded transaction can be found in the hex field of the object (if present) (e.g. as a result of signrawtransactionwithwallet).
+
+	// plain hex transaction
+
+	#[test]
+	fn into_tx_from_json_response_with_invalid_hex_data() {
+		let response = JsonResponse(serde_json::json!("foobar"));
+		match TryInto::<Transaction>::try_into(response) {
+			Err(e) => {
+				assert_eq!(e.kind(), std::io::ErrorKind::InvalidData);
+				assert_eq!(e.get_ref().unwrap().to_string(), "invalid hex data");
+			}
+			Ok(_) => panic!("Expected error"),
+		}
+	}
+
+	#[test]
+	fn into_tx_from_json_response_with_invalid_data_type() {
+		let response = JsonResponse(Value::Number(Number::from_f64(1.0).unwrap()));
+		match TryInto::<Transaction>::try_into(response) {
+			Err(e) => {
+				assert_eq!(e.kind(), std::io::ErrorKind::InvalidData);
+				assert_eq!(e.get_ref().unwrap().to_string(), "expected JSON string");
+			}
+			Ok(_) => panic!("Expected error"),
+		}
+	}
+
+	#[test]
+	fn into_tx_from_json_response_with_invalid_tx_data() {
+		let response = JsonResponse(serde_json::json!("abcd"));
+		match TryInto::<Transaction>::try_into(response) {
+			Err(e) => {
+				assert_eq!(e.kind(), std::io::ErrorKind::InvalidData);
+				assert_eq!(e.get_ref().unwrap().to_string(), "invalid transaction");
+			}
+			Ok(_) => panic!("Expected error"),
+		}
+	}
+
+	#[test]
+	fn into_tx_from_json_response_with_valid_tx_data() {
+		let tx_bytes = Vec::from_hex("0100000001a15d57094aa7a21a28cb20b59aab8fc7d1149a3bdbcddba9c622e4f5f6a99ece010000006c493046022100f93bb0e7d8db7bd46e40132d1f8242026e045f03a0efe71bbb8e3f475e970d790221009337cd7f1f929f00cc6ff01f03729b069a7c21b59b1736ddfee5db5946c5da8c0121033b9b137ee87d5a812d6f506efdd37f0affa7ffc310711c06c7f3e097c9447c52ffffffff0100e1f505000000001976a9140389035a9225b3839e2bbf32d826a1e222031fd888ac00000000").unwrap();
+		let target_tx = deserialize(&tx_bytes).unwrap();
+		let response = JsonResponse(serde_json::json!(encode::serialize_hex(&target_tx)));
```

Comment:
Not quite sure I follow the conversions. Is the string literal on line 655 the same as the hex produced by `encode::serialize_hex` on line 657? Would it be simpler to take say the coinbase `Transaction` from the genesis block as expected and converted to hex for the JSON response.

Reply:
Yes, indeed.